homeassistant/components/tellduslive/__init__.py

```diff
@@ -4,250 +4,160 @@
-        _LOGGER.debug('Continuing setup configured by configurator')
-    elif conf and CONF_HOST in next(iter(conf.values())):
-        #  For now, only one local device is supported
-        _LOGGER.debug('Using Local API pre-configured by configurator')
-        session = Session(**next(iter(conf.values())))
-    elif DOMAIN in conf:
-        _LOGGER.debug('Using TelldusLive cloud service '
-                      'pre-configured by configurator')
-        session = Session(PUBLIC_KEY, NOT_SO_PRIVATE_KEY,
-                          application=APPLICATION_NAME, **conf[DOMAIN])
-    elif config.get(DOMAIN):
-        _LOGGER.info('Found entry in configuration.yaml. '
-                     'Requesting TelldusLive cloud service configuration')
-        request_configuration()
-
-        if CONF_HOST in config.get(DOMAIN, {}):
-            _LOGGER.info('Found TelldusLive host entry in configuration.yaml. '
-                         'Requesting Telldus Local API configuration')
-            request_configuration(config.get(DOMAIN).get(CONF_HOST))
-
-        return True
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN:
+        vol.Schema({
+            vol.Optional(CONF_HOST, default=DOMAIN):
+            cv.string,
+            vol.Optional(CONF_UPDATE_INTERVAL, default=SCAN_INTERVAL):
+            (vol.All(cv.time_period, vol.Clamp(min=MIN_UPDATE_INTERVAL)))
+        }),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+DATA_CONFIG_ENTRY_LOCK = 'tellduslive_config_entry_lock'
+CONFIG_ENTRY_IS_SETUP = 'telldus_config_entry_is_setup'
+
+
+async def async_setup_entry(hass, entry):
+    """Create a tellduslive session."""
+    from tellduslive import Session
+    conf = entry.data
+
+    if KEY_HOST in conf:
+        session = Session(**conf)
     else:
-        _LOGGER.info('Tellstick discovered, awaiting discovery callback')
-        return True
+        session = Session(
+            PUBLIC_KEY,
+            NOT_SO_PRIVATE_KEY,
+            application=APPLICATION_NAME,
+            **conf,
+        )
 
     if not session.is_authorized:
         _LOGGER.error('Authentication Error')
         return False
 
-    client = TelldusLiveClient(hass, config, session)
+    hass.data[DATA_CONFIG_ENTRY_LOCK] = asyncio.Lock()
+    hass.data[CONFIG_ENTRY_IS_SETUP] = set()
+
+    client = TelldusLiveClient(hass, entry, session)
     hass.data[DOMAIN] = client
-    client.update()
 
-    interval = config.get(DOMAIN, {}).get(CONF_UPDATE_INTERVAL,
-                                          DEFAULT_UPDATE_INTERVAL)
+    await client.update()
+
+    interval = hass.data.get("{}_{}".format(DOMAIN, CONF_UPDATE_INTERVAL),
+                             SCAN_INTERVAL)
     _LOGGER.debug('Update interval %s', interval)
-    track_time_interval(hass, client.update, interval)
+    async_track_time_interval(hass, client.update, interval)
+
+    return True
+
 
+async def async_setup(hass, config):
+    """Set up the Telldus Live component."""
+    hass.async_create_task(
```

Comment:
If we set up via user config entry, ie with no config section, we should just `return True` early here.

Reply:
This is for backwards compatibility with old config.

Reply:
Yes. But this is not the only case that can happen. We will always set up the component, but only the import case will use the config. The other cases should just return `True`.

Reply:
I thought that the component only would be called if it had entries in configuration.yaml.

Reply:
No, the component is required to be set up always. Here are the different cases, besides from config yaml for config entry supporting components.
- After entry creation: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/config_entries.py#L498
- During entry forwarding: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/config_entries.py#L451
- Existing stored created entries during bootstrap/setup: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/bootstrap.py#L124
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/bootstrap.py#L149